src/shared/utils/project-utils.ts

```diff
@@ -161,3 +193,136 @@ export const createPackageJSON = (
+
+  return createFile('package', FILE_EXTENSIONS.JSON, JSON.stringify(content, null, 2))
+}
+
+export const createPageFile = async (params: PageFactoryParams): Promise<GeneratedProjectData> => {
+  const {
+    componentGenerator,
+    stateBranch,
+    routerDefinitions,
+    componentOptions: options,
+    pageMetadataOptions,
+    pageExtension,
+  } = params
+
+  const files: GeneratedFile[] = []
+  let dependencies: Record<string, string> = {}
+  const { value, content } = stateBranch
+
+  if (typeof value !== 'string' || typeof content === 'string' || !routerDefinitions) {
+    return { files, dependencies }
+  }
+
+  const { componentName, fileName } = extractPageMetadata(routerDefinitions, value, {
+    ...pageMetadataOptions,
+  })
+  const pageUIDL = createComponentUIDL(componentName, content, { fileName })
+
+  try {
+    const compiledPageComponent = await componentGenerator.generateComponent(pageUIDL, {
+      ...options,
+    })
+    const { externalCSS, externalDependencies, code } = compiledPageComponent
+    dependencies = externalDependencies
+
+    if (externalCSS) {
+      const cssFile = createFile(fileName, FILE_EXTENSIONS.CSS, externalCSS)
+      files.push(cssFile)
+    }
+
+    const fileExtension = pageExtension || FILE_EXTENSIONS.JS
+    const pageFile = createFile(fileName, fileExtension, code)
+    files.push(pageFile)
+  } catch (error) {
+    console.warn(componentName, error)
+  }
+
+  return { files, dependencies }
+}
+
+export const createComponentFile = async (
+  params: ComponentFactoryParams
+): Promise<GeneratedProjectData> => {
+  let dependencies: Record<string, string> = {}
+  const files: GeneratedFile[] = []
+  const { componentGenerator, componentUIDL, componentExtension, componentOptions } = params
+
+  try {
+    const compiledComponent = await componentGenerator.generateComponent(componentUIDL, {
+      ...componentOptions,
+    })
+
+    const { code, externalCSS, externalDependencies } = compiledComponent
+    const fileName = sanitizeVariableName(componentUIDL.name)
+    dependencies = externalDependencies
+
+    if (externalCSS) {
+      const cssFile = createFile(fileName, FILE_EXTENSIONS.CSS, externalCSS)
+      files.push(cssFile)
+    }
+
+    const fileExtension = componentExtension || FILE_EXTENSIONS.JS
+    const componentFile = createFile(fileName, fileExtension, code)
+    files.push(componentFile)
+  } catch (error) {
+    console.warn(componentUIDL.name, error)
+  }
+  return { files, dependencies }
+}
+
+function createComponentUIDL(
+  name: string,
+  content: ContentNode,
+  meta: Record<string, any>
+): ComponentUIDL {
+  return {
+    name,
+    content,
+    meta: {
+      ...meta,
+    },
+  }
+}
+
+export const joinComponentFiles = (
+  componentFiles: GeneratedProjectData[]
+): GeneratedProjectData => {
+  let dependencies = {}
+
+  const files = componentFiles.reduce((componentFile: GeneratedFile[], currentFile) => {
```

Comment:
you can reduce over a { dependencies, files } accumulator so you can do both concats in one assignment, without needing to separate dependencies as a side effect in each of the reduce functions.

also the name of the function is misleading, since it also joins the dependencies.